### ISSUE TITLE
CFP: Investigate if it's possible to support GatewayAddress in Gateway API

### ISSUE THREAD

**Issue by sayboras** [cilium/cilium#21926]
## Cilium Feature Proposal
        
The first cut for Gateway API was done in https://github.com/cilium/cilium/pull/21749, however, we have a couple of follow-up tasks to support more features from Gateway API upstream.

This issue is to track the work of supporting GatewayAddress in Gateway resources. The current design is to just offload this to Cloud Provider by creating LoadBalancer service type i.e. once the LB service is provisioned and ready, we just take the allocated IP/Hostname from the LoadBalancer service and propagate the same to related Gateway. In order to support _GatewayAddress_, we need to investigate if similar capability is possible in different cloud providers (e.g. AWS, Google cloud, Azure to start with). My hunch is to just add the [annotation](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) in LB service. 

- [ ] Gateway 
   - [ ] Support for Spec.Address (if possible). 
   
## Goals

Investigate if it's possible with the current design, if not, we should make it clear in the documentation.

## Proposed solution
TBD

## Tasks

TBD

**Comment by dylandreimerink** [comment]
My two cents on the issue. Services have the `.spec.loadBalancerIP` field which can be used to request a single IP and is universally supported. Now this field is deprecated according to [comments on the go struct](https://github.com/kubernetes/api/blob/v0.25.4/core/v1/types.go#L4464). 

> // As of Kubernetes v1.24, users are encouraged to use implementation-specific annotations when available.
> // This field may be removed in a future API version.

In the case of the LB-IPAM feature, we have a fairly straightforward `io.cilium/lb-ipam-ips` annotations. Cloud providers tend to have a list of annotations which can be used to together, not only including a requested IP but also a lot of other settings:
* [AWS](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/)
* [Azure/AKS](https://learn.microsoft.com/en-us/azure/aks/load-balancer-standard#additional-customizations-via-kubernetes-annotations)
* [GKE](https://cloud.google.com/kubernetes-engine/docs/how-to/load-balance-ingress#summary_of_external_ingress_annotations)

The easiest and most straightforward route would be to somehow forward annotations from the Gateway object to the Service object. However, I have no clue if there is anything in the Gateway API spec at this time that allows that.

The most straightforward way within the current spec would be to just translate the requested addresses into all of the well-known annotations for the different providers. You would miss out on the "additional" settings which do not fit the IP or Hostname primitive.

My last suggestion, which would be way more work, would be to introduce [custom address types](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.AddressType) for every LB provider we want to support. Then allow users to specify all of their settings in the value which can be a K/V format and translate these to the proper annotations on the service.

For example:
`aws.amazon.com/alb: load-balancer-name=abc,ip-address-type=ipv4,shield-advanced-protection=true`
Would translate to
`alb.ingress.kubernetes.io/load-balancer-name: abc`, `alb.ingress.kubernetes.io/ip-address-type: ipv4`, `alb.ingress.kubernetes.io/shield-advanced-protection: true`

The obvious downside is that we now have to maintain this translation, keep-up-to-date with new annotations added and so forth. Though I see this at least being an option for our own LB-IPAM.

**Comment by youngnick** [comment]
One of the major goals for Gateway API is to remove this sort of annotation proliferation, so I'm pretty strongly -1 on anything that copies annotations around. I think that I'd like to see something like this:
- We maintain a mapping of "Gateway.spec.addresses entries equals this annotation on each could provider"
- Other Loadbalancer settings are set in a GatewayClass ParametersRef (so a cluster can have a GatewayClass that allows shield advanced protection, and one that doesn't by having different config in the ParametersRef).

For AWS, choosing ELB/NLB/ALB should probably be handled by different GatewayClasses, and then you create Gateways that create a LB of that type.

#21923 covers the work to build out Cilium support for ParametersRef.

**Comment by youngnick** [comment]
We'll need to do this to finish out #21929.

**Comment by j-gooding** [comment]
I don't see any mention of on-perm deployments in this thread. How would this work with BGP Control Plane, or L2 Announcements (v1.14) for those who are not in cloud based situations? 

I'm currently using MetalLb since cilium LB IPAM doesn't properly support all CIDRs or arbitrary ranges such as pools of one (https://github.com/cilium/cilium/issues/24351).

Also consider the scenario where someone needs the Gateway address / load balancer to be the _only_ public address, and the backend services are all of type cluster IP. For example your gateway is a single A + PTR address/record, and then all of the services are CNAME record subdomains which point to the A record, and the gateway breaks it up by host, eg wiki.mysub.university.edu, log.mysub.university.edu, dash.mysub.university.edu, etc.

I think if cilium is going to make a point to try and support smaller on perm deployments as specifically stated in the L2 announcements documentation, then these kind of things should be carefully considered and also documented.

**Comment by youngnick** [comment]
Thanks for that @j-gooding. I think the asks you have there are reasonable, and we should support them when we can.

In terms of the Gateway API support, when we've been thinking about #21929, it seems likely that we'll end up with something like this:
- For the shorter term, we'll create LB Services to fulfill a Gateway request in as many places as we can, and ensure that those LB services will allow setting the address.
- This will also apply for Cilium's LBIPAM.
- For other relevant annotations or LB Service config, we'll probably look to make use of the new `infrastructure` stanza that's being added to Gateway (see kubernetes-sigs/gateway-api#1868 for the initial PR) - annotations are a likely use case for that stanza too.

We also need to add the code to actually look at `spec.addresses` for Gateways - currently the Cilium code basically ignores this stanza altogether. This shouldn't be too hard, but we've missed the window for 1.14 features (feature freeze started last Friday), so we'll need to see if this meets the backport criteria once it's done, otherwise, it should hopefully be in 1.15.

**Comment by lukapetrovic-git** [comment]
> I don't see any mention of on-perm deployments in this thread. How would this work with BGP Control Plane, or L2 Announcements (v1.14) for those who are not in cloud based situations?
> 
> I'm currently using MetalLb since cilium LB IPAM doesn't properly support all CIDRs or arbitrary ranges such as pools of one (#24351).
> 
> Also consider the scenario where someone needs the Gateway address / load balancer to be the _only_ public address, and the backend services are all of type cluster IP. For example your gateway is a single A + PTR address/record, and then all of the services are CNAME record subdomains which point to the A record, and the gateway breaks it up by host, eg wiki.mysub.university.edu, log.mysub.university.edu, dash.mysub.university.edu, etc.
> 
> I think if cilium is going to make a point to try and support smaller on perm deployments as specifically stated in the L2 announcements documentation, then these kind of things should be carefully considered and also documented.

I agree with this 100%. Currently in the same situation on-prem with BGP Control Plane, would make a lot of sense to me to have this functionality.

**Comment by ksully-execute** [comment]
Might this be on the plan for upcoming releases?  or not a priority since it's more about on-prem deployments?

**Comment by youngnick** [comment]
Yes, we would like to get this done, just a matter of picking an approach, and deciding how closely this will work with BGP control plane and L2Announcment.

**Comment by larssb** [comment]
This is not solely for on-prem envs. It can be very important to control IP Address assignments in the cloud as well. E.g. if you intend the gateway **NOT** to be exposed publicly. Then you want an internal type LoadBalancer service in Kubernetes ... and that being the case ( at least that's how it is in GCP ), you have to specify an IP that the cloud-provider provisioned LB gets.

In certain cases it might also be beneficial to control the public ip that an GW API instance gets. E.g. if your company have it's on public range of IP's.

Thanks.

**Comment by youngnick** [comment]
Thanks for that @larssb. I agree that it's important to also support cloud environments, but for those, the best option is probably going to be support for creating an LB Service, as I mentioned above.

However, it turns out that there's no standard way to request an address on cloud providers - all of the big 3 require annotations to be set on the generated Service, so to support this, we'll need to special-case `spec.address` support per cloud-provider. Which sucks. I'm going to discuss this upstream with some SIG-Network folks and see if there is anything planned for Service to help with that. (I suspect the answer will be "no, that's what Gateway is for", which leaves us in a lovely paradox.)

All of this is to explain why this is not done already - we'll see if we can come up with anything else.

**Comment by youngnick** [comment]
I should note that because all of the cloud providers support setting IP addresses etc via annotations, you can currently get around this limitation using the `infrastructure` stanza in Gateway, implemented in #25357.

That's in no way the best solution here, but for anyone finding this issue in the meantime, may keep you moving while we sort this out. Again, sorry to not be able to just solve this one.

**Comment by larssb** [comment]
Hi @youngnick,

Thank you very much for chiming in with some great details.

Hmm I can't seem to find a proper annotation to use on `Kind: Service` of the `LoadBalancer` type on Google Cloud. The only annotation coming close to making sense is `kubernetes.io/ingress.global-static-ip-name` found at [](https://cloud.google.com/kubernetes-engine/docs/tutorials/configuring-domain-name-static-ip#use_an_ingress). However, as I understand it, this annotation having the following downsides.

- It requires extra actions. E.g. the reservation of the IP on Google Cloud
- I'm not even sure that it works with an `Gateway API Gateway` instance

But, yes, the `networking.gke.io/load-balancer-type: "Internal"` could be set via the `.spec.infrastructure.annotations` stanza. We just seems to end up in a half-baked situation where I still have to manipulate the underlying generated `Kind: Service` of the `LoadBalancer` type.

---

Maybe I should also try/use the `Kyverno` route of using a `MutatingWebhookConfiguration` like others. Or potentially I've misunderstood or just didn't find the proper annotation to use.

Thanks.

**Comment by youngnick** [comment]
Hmm, setting the Service `.spec.loadBalancerIP` may be a reasonable first step for Cilium to do that would add _some_ support for Gateway `spec.addresse`. I'll look into that, thanks @larssb!

**Comment by larssb** [comment]
I'm happy to hear that you back that part @youngnick. Much appreciated. I think too ( unsurprisingly :-) ). Also `spec.address`is not cloud provider specific - obviously - so should be a okay to use and support.

Thank you very much.

**Comment by larssb** [comment]
Aaaah good point @larivierec!

I wasn't properly aware on the deprecation of the `.spec.loadBalancerIP `. Important to know. But, as mentioned in the ticket/issue you refer to: "it'll properly take along time before that field is fully truly deprecated and removed entirely". So far no cigar at least.

But, then it would be great that `Cilium's Gateway API` implementation supported the plural field in the spec on `Kind: Service` of the `LoadBalancer` type. Because, as previously stated the `annotation` option seems to differentiate a great deal between cloud solutions and e.g. GCP ( at least according to what I've been able to identify ), do not support controlling the IP in a good way through an `service annotation`.

Thanks.

**Comment by rouke-broersma** [comment]
> Thanks for that @larssb. I agree that it's important to also support cloud environments, but for those, the best option is probably going to be support for creating an LB Service, as I mentioned above.
> 
> However, it turns out that there's no standard way to request an address on cloud providers - all of the big 3 require annotations to be set on the generated Service, so to support this, we'll need to special-case `spec.address` support per cloud-provider. Which sucks. I'm going to discuss this upstream with some SIG-Network folks and see if there is anything planned for Service to help with that. (I suspect the answer will be "no, that's what Gateway is for", which leaves us in a lovely paradox.)
> 
> All of this is to explain why this is not done already - we'll see if we can come up with anything else.

Tbh this seems like a flaw in the spec design and in my opinion I would prefer it get dropped entirely. 

We already have types for requesting loadbalancing. The spec is suggesting that a valid implementation could be that during gateway creating cloud APIs are used to request the addresses, ie cloud controllers need to implement Gateway to watch for requested addresses, or the gateway controller needs to implement cloud provider APIs. 

A better way to me would be for the infrastructure stanza to allow selecting an existing loadbalancer service as part of the infrastructure request if specific addresses are necessary.

I really wonder if they have thought of a reference implementation for this feature, because putting all this responsibility on every Gateway controller implementation seems like a big ask to me.

Especially since service.spec.loadbalancerip has been deprecated because they felt that the configuration of external loadbalancers is too cloud specific and thus does not fit in the core spec, I don't understand how they more or less decided to bring it back with Gateway Api.

**Comment by rouke-broersma** [comment]
I recently came across the azure implementation for gateway api and this is very in line with my issues with the current spec:

https://learn.microsoft.com/en-us/azure/application-gateway/for-containers/how-to-ssl-offloading-gateway-api?tabs=byo

They gateway api implementation uses a custom controller that handles everything that is needed of the cloud provider. This is very nice, if you're using azure. And this works perfectly with the cloud-native gateway api implementation, They can simply take the requested ip from the gateway, provision this in azure and light up the integration in kubernetes. They don't even need to use a Service to request the load balancer, they can do it directly. Any cross-cloud or more generic gateway api implementation can't do this without having to build support for every cloud provider api. This new gateway api spec seems very cloud provider focused, and not so friendly to cross-cloud vendors.

**Comment by youngnick** [comment]
The intent with Gateway API is for it to be widely implemented, yes. And ideally, eventually the cloud providers will support Gateway resources with TCPRoutes or UDPRoutes as an alternative to Loadbalancer Services. (Cilium's LBIPAM support will probably grow direct support like this eventually too).

The Gateway API is explicitly designed to be used _both_ by cloud providers in this way, and north/south use cases (replacing an Ingress controller).

There has been some discussion of selecting existing LB Services as an option in `infrastructure`, but there has been no consensus about it.

I'd suggest bringing your feedback to the Gateway API project itself, we're always looking for more passionate users.

**Comment by chrede88** [comment]
What's the current status of this? I'm using L2 and Gateway API and would really like to be able to request a specific IP from my pool. Right now, I've to make sure my two gateways comes up in the "correct" order so they get the "correct" IP, which really feel like a horrible hack.

**Comment by tchellomello** [comment]
+1 here. We would love to have this ability!

**Comment by xaviergrandmougin** [comment]
I would be thrilled if this would be implemented!

For now,as a workaround, I manually edit the service created by the gateway, and add the `io.cilium/lb-ipam-ips` annotation to get the IP I want for the pool.

I would be delighted to get rid of this ugly hack, and would be happy if you can tell a less ugly way to do it in the meantime.

**Comment by rouke-broersma** [comment]
> I would be thrilled if this would be implemented!
> 
> For now, as a workaround, I manually edit the service created by the gateway, and add the `io.cilium/lb-ipam-ips` annotation to get the IP I want from the pool.
> 
> I would be delighted to get rid of this ugly hack, and would be happy if you can tell a less ugly way to do it in the meantime.

You can do this with the experimental crd and the infrastructure stanza already.

**Comment by ksully-execute** [comment]
@rouke-broersma  Can you point us at the docs for the crd and infrastructure stanza?

**Comment by rouke-broersma** [comment]
> @rouke-broersma  Can you point us at the docs for the crd and infrastructure stanza?

https://github.com/cilium/cilium/issues/21926#issuecomment-1882556513

**Comment by youngnick** [comment]
I'd definitely like to implement this functionality, but it will need to work differently on each platform:
- Cilium lb-ipam
- Cilium BGP
- Cloud providers require specific annotations or labels on the created Service

Folks who are interested here, could I please ask that you comment with the environment you want to be able to set an address in? That will let us prioritize which work we do first.

**Comment by rouke-broersma** [comment]
Azure
BGP

**Comment by jackmaninov** [comment]
And I'm cilium LB-IPAM, though easily surviving with IPs requested through the infrastructure stanza for the moment. Although that should make it trivial to support 😉

**Comment by lukapetrovic-git** [comment]
> I'd definitely like to implement this functionality, but it will need to work differently on each platform:
> 
> * Cilium lb-ipam
> * Cilium BGP
> * Cloud providers require specific annotations or labels on the created Service
> 
> Folks who are interested here, could I please ask that you comment with the environment you want to be able to set an address in? That will let us prioritize which work we do first.

Hi, i am using:
- lb-ipam
- BGP
- on-prem environment

**Comment by xaviergrandmougin** [comment]
I'm using it on:
- LB-IPAM with L2 announcements
- on-prem setup with kubeadm and kubeProxyReplacement

**Comment by ksully-execute** [comment]
for me:


    LB-IPAM with L2 announcements
    on-prem setup with kubeadm and kubeProxyReplacement

**Comment by jorhett** [comment]
Same here:
> LB-IPAM with L2 announcements
> on-prem setup with kubeadm and kubeProxyReplacement

I really wish you could tag comments with +1

**Comment by ksully-execute** [comment]
Pardon my ignorance,  how and why?From: Jo Rhett ***@***.***>Sent: Wednesday, July 17, 2024 2:30:08 PMTo: cilium/cilium ***@***.***>Cc: ksully-execute ***@***.***>; Comment ***@***.***>Subject: Re: [cilium/cilium] CFP: Investigate if it's possible to support GatewayAddress in Gateway API (Issue #21926)
Same here:

LB-IPAM with L2 announcements
on-prem setup with kubeadm and kubeProxyReplacement

I really wish you could tag comments with +1

**Comment by youngnick** [comment]
@chaunceyjiang is proceeding on work on #33042 to add this functionality for LB-IPAM only (for now). Once that gets in, we'll see about extending the functionality to other platforms.

I encourage anyone who finds this issue and is interested in their platform being supported to comment with their platform. Notable ones are Cilium's BGP support, or the various cloud provider Loadbalancer Services (like GKE, AKS, and EKS).

**Comment by alessio-pascolini** [comment]
```
on-prem
kube-proxy replacement
LB-IPAM
cilium BGP Control Plane
```

**Comment by blufor** [comment]
> I'd definitely like to implement this functionality, but it will need to work differently on each platform:
> 
> * Cilium lb-ipam
> * Cilium BGP
> * Cloud providers require specific annotations or labels on the created Service
> 
> Folks who are interested here, could I please ask that you comment with the environment you want to be able to set an address in? That will let us prioritize which work we do first.

- on-prem (k3s)
- kube-proxy-less
- LB-IPAM
- BGP Control Plane (for OpenBSD firewall in front of the cluster that runs bgpd and pf)

**Comment by ksully-execute** [comment]
On-prem

**Comment by SixFive7** [comment]
On-premise
Cilium instead of kube-proxy
LB IPAM
L2 arp with plans for BGP

**Comment by figaw** [comment]
- on-prem
- kube-proxy-less
- lb-ipam
- Cilium BGP

**Comment by youngnick** [comment]
Thanks very much for all the feedback folks! With #33042 in review, we are a fair way along at supporting LB-IPAM, so once that is merged, I'll note here and everyone can test.

I'd still like to encourage people who _aren't_ using LB-IPAM to comment here if you need Gateway Address support for your platform though!

**Comment by vehagn** [comment]
@youngnick Cilium supports the Gateway `addresses` field now, so I suppose this issue can be closed?

It works for me at least!

**Comment by xaviergrandmougin** [comment]
I confirm, it works for me too.

**Comment by youngnick** [comment]
Great, thanks everyone. I'll leave this open so that people can ask if they want other Loadbalancer Service providers supported though.

**Comment by youngnick** [comment]
Hmm, I think at this point, I'm going to declare this one done. Cilium does support Gateway `addresses` field, as long as you're using Cilium's LB-IPAM.
